Addons/tsgExoticQuadrature.hpp

```diff
@@ -45,30 +45,40 @@ inline double poly_eval(const std::vector<double> &roots, double x) {
     return eval_value;
 }
 
-//! \brief Generate the roots of all orthogonal polynomials up to degree n, where orthogonality is with respect to the measure
-//! given by ref_weights. It is assumed that the values of ref_weights are nonnegative.
+//! \brief Generate the roots of all orthogonal polynomials up to degree n, where orthogonality is with respect to the
+// measure given by ref_weights. Specializes depending on if the underlying weight function is symmetric or not.
+template <bool is_symmetric>
 inline std::vector<std::vector<double>> getRoots(const int n, const std::vector<double> &ref_weights,
                                                  const std::vector<double> &ref_points) {
 
     // Compute the roots incrementally.
     assert(ref_points.size() == ref_weights.size());
     assert(ref_points.size() > 0);
+    if (not std::all_of(ref_weights.begin(), ref_weights.end(), [](double x){return x>=0.0;})) {
+        throw std::invalid_argument("ref_weights needs to be a nonnegative vector!");
+    }
     std::vector<double> poly_m1_vals(ref_points.size(), 0.0), poly_vals(ref_points.size(), 1.0);
     std::vector<double> diag, offdiag;
     std::vector<std::vector<double>> roots(n);
     for (int i=0; i<n; i++) {
-        // Form the tridiagonal vectors, alpha and beta, of the Jacobi matrix.
-        double diag_numr = 0.0;
-        double diag_denm = 0.0;
+        // Form the diagonal and offdiagonal vectors of the Jacobi matrix.
+        if (is_symmetric) {
+            diag.push_back(0.0);
+        } else {
+            double diag_numr = 0.0;
+            double diag_denm = 0.0;
+            for (size_t j=0; j<ref_points.size(); j++) {
+            diag_numr += ref_points[j] * (poly_vals[j] * poly_vals[j]) * ref_weights[j];
+            diag_denm += (poly_vals[j] * poly_vals[j]) * ref_weights[j];
+            }
+            diag.push_back(diag_numr / diag_denm);
+        }
         double sqr_offdiag_numr = 0.0;
         double sqr_offdiag_denm = 0.0;
         for (size_t j=0; j<ref_points.size(); j++) {
-            diag_numr += ref_points[j] * (poly_vals[j] * poly_vals[j]) * ref_weights[j];
-            diag_denm += (poly_vals[j] * poly_vals[j]) * ref_weights[j];
             sqr_offdiag_numr += (poly_vals[j] * poly_vals[j]) * ref_weights[j];
             sqr_offdiag_denm += (poly_m1_vals[j] * poly_m1_vals[j])  * ref_weights[j];
         }
-        diag.push_back(diag_numr / diag_denm);
         if (i >= 1) {
```

Comment:
Do not execute the loop and compute `sqr_offdiag_numr` and `sqr_offdiag_denm` unless you will use them in the `if (i >= 1)` case. Basically, put the loop inside the if-statement.

Reply:
Done.